netket/_vmc.py

```diff
@@ -145,19 +148,20 @@ def _forward_and_backward(self):
             self._grads, self._jac = self._machine.vector_jacobian_prod(
                 samples_r, eloc_r / self._n_samples, self._grads, return_jacobian=True
             )
-            _sum_inplace(self._grads)
-
-            # Center the log derivatives
-            self._jac -= _mean(self._jac, axis=0)
-
+            
+            self._grads = tree_map(_sum_inplace, self._grads)
+            self._grads, self._jac = shape_for_sr(self._grads,self._jac)
```

Comment:
Would it be possible to adsorb these two function calls ```shape_for_sr``` and ```shape_for_update``` directly into the Jax SR solver? 

This way, you could also remove any reference to Jax in this file, i.e. the lines 
```python
from netket.utils import jax_available
if jax_available:
    from netket.vmc_common import shape_for_sr, shape_for_update
```

Reply:
This would be tricky because SR doesn't have access to the shape of the parameter tree. I guess I could have flatten return the shape and use it for unflatten

Reply:
but in the Jax SR you can just use the native ```flatten``` and ```unflatten``` from Jax that should allow you to pack and unpack at will! Actually I was just looking at what Jax does with optimizers, and basically it seems that they convert these trees into flat arrays, if I understand correctly [this](https://github.com/google/jax/blob/f60184e12e279c4602c77f91be19a0c6e1eb7083/jax/experimental/optimizers.py#L100)

Reply:
the native things are 
```python
from jax.tree_util import (tree_map, tree_flatten, tree_unflatten)
```

Reply:
Yeah, for some reason tree_flatten doesn't actually return a flat arrays, all it seems to do is concatenate all of the parameters into a list instead of nested lists (i.e. before flatten the params of each layer are in their own branch). That's why numpy_flatten in "netket/machine/jax.py" still has to iterate even after calling tree_flatten. I was also confused by this

Reply:
But it seems to me that you can still easily move these two 

```python 
shape_for_sr(self._grads,self._jac)

shape_for_update(self._dp,self._machine.parameters) 
```

in the JaxSr class. In general the rationale is that the JaxSr class is supposed to work with Jax types, and that the Sr class we have now is supposed to work only with numpy types. Later on we will think about an automatic dispatch system...

Reply:
Okay, I did this. Since SR doesn't have access to "self._machine.parameters", shape_for_sr now returns a copy of grads in order to shape back to Pytree. Since grads consumes. memory 1/batch size compared to Jac this didn't seem like a terrible solution but let me know if there's something better. I would return just the shapes, but tree_unflatten needs an object "shape_like"

Reply:
yes I see now the issue
Then the solution is just to construct Sr passing the machine explicitly in its __init__ or in vmc there is these lines 
```python
        if sr is not None:
            self._sr.is_holomorphic = sampler.machine.is_holomorphic
```
where you can add something like 

```python
        if sr is not None:
            self._sr.is_holomorphic = sampler.machine.is_holomorphic
            self._sr.machine=sampler.machine 
```

Reply:
Okay great, this works.